### PR TITLE
return a promise if no callback is passed

### PR DESCRIPTION
**Pull request by peterkhayes**
This will allow us to use async/await more extensively in our code.

I had to fix some eslint stuff - JP had apparently left this in a state where linting didn't pass.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -2,12 +2,16 @@ module.exports = {
   extends: ['prettier', 'airbnb-base/legacy'],
   plugins: ['prettier'],
   rules: {
-    'prettier/prettier': ['error', { singleQuote: true, printWidth: 100 }],
+    'prettier/prettier': ['warn', { singleQuote: true, printWidth: 100 }],
```
Comment (rf-) [inline .eslintrc.js]: 😱 Did you leave this in by accident?

File: index.js [hunk 2]
```diff
@@ -216,4 +229,6 @@ module.exports = function newXhrQueue(options) {
       refreshQueue(true);
     }
   };
+
+  return result;
```
Comment (rf-) [inline index.js]: Also was this change intentional?

File: index_spec.js [hunk 3]
```diff
@@ -406,4 +406,11 @@ describe('newXhrQueue', function() {
       expect(callbackForSuccessfulRequest).toHaveBeenCalled();
     });
   });
+
+  it('returns a promise if no callback is passed', function() {
+    var promise = xhrQueue.xhr({ url: '/write-1', method: 'POST' });
+    jasmine.clock().tick(1);
+    respondToRequestWithUrl('/write-1');
+    return promise;
+  });
```
Comment (rf-) [inline index_spec.js]: It seems kind of strange that this doesn't assert anything or examine the promise at all

**Comment by janpaul123** [comment]
> JP had apparently left this in a state where linting didn't pass.

Oopsies!

File: index_spec.js [hunk 4]
```diff
@@ -406,4 +406,11 @@ describe('newXhrQueue', function() {
       expect(callbackForSuccessfulRequest).toHaveBeenCalled();
     });
   });
+
+  it('returns a promise if no callback is passed', function() {
+    var promise = xhrQueue.xhr({ url: '/write-1', method: 'POST' });
+    jasmine.clock().tick(1);
+    respondToRequestWithUrl('/write-1');
+    return promise;
```
Comment (rf-) [inline index_spec.js]: Is there an assertion we could make here to more clearly test that a promise is returned?

File: .eslintrc.js (hunk 1 above)
Comment (rf-) [inline .eslintrc.js]: Shouldn’t this be `error` for the same reasons as the monorepo?

File: index.js [hunk 5]
```diff
@@ -153,8 +155,19 @@ module.exports = function newXhrQueue(options) {
...
+  var result = {
     xhr: function(request, callback) {
+      if (arguments.length === 1) {
+        return new Promise(function(resolve, reject) {
+          result.xhr(request, function(err, response) {
+            if (err) {
+              return reject(err);
+            }
+            return resolve(response);
+          });
```
Comment (rf-) [inline index.js:167]: It's kind of a bummer that we're losing the third callback param (`body`) here. Maybe we could add it to `response` as a property or something?

File: index_spec.js (hunk 4 above)
Comment (peterkhayes) [reply index_spec.js]: Hmm, good idea.  I added an assertion for this.

File: index.js (hunk 5 above)
Comment (peterkhayes) [reply index.js:167]: it's already there! `response.body` should be valid.

File: .eslintrc.js (hunk 1 above)
Comment (peterkhayes) [reply .eslintrc.js]: I made it a warn because for awhile I couldn't get it all to pass somehow.  But now I changed it back and it should work.

**Review by rf- — APPROVED** [review]

**Comment by peterkhayes** [comment]
@janpaul123 can you publish this, or better, add me as a collaborator on NPM?

**Comment by janpaul123** [comment]
Added you to all remix-based projects on npm along with @tilgovi and @rf- 😄
Thanks for the improvement!!

**Comment by harto** [comment]
I just discovered this PR! Looks great. I just realized that this API doesn't have the ability to cancel requests (the callback-based API returns an object `{ cancel: () => {…} }`). I wonder if we want to preserve that functionality in the Promise-based API?

**Comment by tilgovi** [comment]
@harto perhaps with this: https://w3ctag.github.io/design-principles/#aborting

**Comment by janpaul123** [comment]
That sounds ideal; or if you want to preserve the old API you can also just add a `cancel` method to the returned promise. #hackyhack 😄
